### PR TITLE
Canvas: Button API Editor visual feedback on response

### LINKED ISSUE #74223: Canvas: Button API Editor Visual Feedback on Response
Add a spinning wheel during API call and show some indication of response, success or failure. Needs to also clear out after an appropriate amount of time or action.

https://github.com/grafana/grafana/assets/22381771/81aa8188-dfcc-4ffd-8a3d-c33a14a92080

### PR DESCRIPTION
**Pull request by adela-almasan**
https://github.com/grafana/grafana/assets/88068998/2a276d06-6787-49e8-b0b3-f15667cb52ba

Fixes #74223

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/plugins/panel/canvas/editor/element/utils.ts [hunk 1]
```diff
@@ -20,15 +23,12 @@ export const callApi = (api: APIEditorConfig, isTest = false) => {
...
-            console.error(error);
-          }
+          appEvents.emit(AppEvents.alertError, ['Error has occurred: ', JSON.stringify(error)]);
+          updateLoadingState && updateLoadingState(false);
         },
         complete: () => {
-          if (isTest) {
-            appEvents.emit(AppEvents.alertSuccess, ['Test successful']);
-          }
+          appEvents.emit(AppEvents.alertSuccess, ['Test successful']);
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/editor/element/utils.ts:30]: I think we might want to update these messages as now it not just the "test" scenario 🤔 
```suggestion
          appEvents.emit(AppEvents.alertSuccess, ['Canvas button API call was successful']);
```

File: public/app/plugins/panel/canvas/editor/element/utils.ts [hunk 2]
```diff
@@ -20,15 +23,12 @@ export const callApi = (api: APIEditorConfig, isTest = false) => {
       .fetch(request)
       .subscribe({
         error: (error) => {
-          if (isTest) {
-            appEvents.emit(AppEvents.alertError, ['Error has occurred: ', JSON.stringify(error)]);
-            console.error(error);
-          }
+          appEvents.emit(AppEvents.alertError, ['Error has occurred: ', JSON.stringify(error)]);
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/editor/element/utils.ts:26]:
```suggestion
          appEvents.emit(AppEvents.alertError, ['An error has occurred: ', JSON.stringify(error)]);
```

File: public/app/features/canvas/elements/button.tsx [hunk 3]
```diff
@@ -40,39 +39,56 @@ export const defaultStyleConfig: ButtonStyleConfig = {
...
-    const onClick = () => {
-      if (data?.api && data?.api?.endpoint) {
-        callApi(data.api);
-      }
-    };
+  const [isLoading, setIsLoading] = React.useState(false);
 
-    return (
-      <Button type="submit" variant={data?.style?.variant} onClick={onClick} className={styles.button}>
+  const updateLoadingState = (loading: boolean) => {
```
Comment (nmarrs) [inline public/app/features/canvas/elements/button.tsx:47]:
```suggestion
  const updateLoadingStateCallback = (loading: boolean) => {
```

maybe?

File: public/app/plugins/panel/canvas/editor/element/utils.ts [hunk 4]
```diff
@@ -7,11 +7,14 @@ import { HttpRequestMethod } from '../../panelcfg.gen';
 
 import { APIEditorConfig } from './APIEditor';
 
-export const callApi = (api: APIEditorConfig, isTest = false) => {
+type IsLoadingFn = (loading: boolean) => void;
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/editor/element/utils.ts:10]:
```suggestion
type IsLoadingCallback = (loading: boolean) => void;
```

File: public/app/plugins/panel/canvas/editor/element/utils.ts [hunk 5]
```diff
@@ -7,11 +7,14 @@ import { HttpRequestMethod } from '../../panelcfg.gen';
...
+
+export const callApi = (api: APIEditorConfig, updateLoadingState?: IsLoadingFn) => {
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/editor/element/utils.ts:12]:
```suggestion
export const callApi = (api: APIEditorConfig, updateLoadingStateCallback?: IsLoadingFn) => {
```

File: public/app/plugins/panel/canvas/editor/element/utils.ts (hunk 1 above)
Comment (nmarrs) [reply public/app/plugins/panel/canvas/editor/element/utils.ts:30]: maybe this doesn't have to be as explicit 🤷‍♂️

**Review by nmarrs — APPROVED** [review]
Just a few nits - functionality looks good overall 💯 


https://github.com/grafana/grafana/assets/22381771/327c6a01-c369-4c37-9def-b04393b239f7

File: public/app/plugins/panel/canvas/editor/element/utils.ts (hunk 1 above)
Comment (adela-almasan) [reply public/app/plugins/panel/canvas/editor/element/utils.ts:30]: oops, missed this. Maybe `API call was successful`?

Comment (nmarrs) [reply public/app/plugins/panel/canvas/editor/element/utils.ts]: sure sounds good 👍

**Pull request merged**